bug/69724 Prevent potential data loss on initial multi-user collaborative document load by deferring editor creation until the provider syncs the client Y.Doc instance

akabiru opened the pull request:
https://community.openproject.org/wp/69724

### Root Cause 

The BlockNote editor was being created and bound to an empty Y.Doc before the Hocuspocus provider synced with the server. This created a potential race condition where auto-save could trigger and send empty updates to the server before receiving the initial document state (sync), possibly causing data loss.

### Approach

Split `OpBlockNoteContainer` into two components using React's component mounting lifecycle as a gate:

- `OpBlockNoteContainer`: Pure provider sync state management. Shows skeleton loader while waiting for provider to sync and only mounts the `BlockNoteEditor` subcomponent after sync completes.

- `BlockNoteEditor`: Encapsulates blocknote editor setup- only mounts when parent confirms sync is done, ensuring `useCreateBlockNote` is called with a Y.Doc that has already received server state.

This leverages React's component lifecycle guarantee: child components only mount when parent renders them.

akabiru committed:
Prevent auto-save race condition by deferring editor creation until provider sync

The BlockNote editor was being created and bound to an empty Y.Doc before
the Hocuspocus provider synced with the server. This created a possible race condition
where auto-save could trigger and send empty updates to the server before
receiving the initial document state, causing data loss.

Split OpBlockNoteContainer into two components using React's
component mounting lifecycle as a gate:

- OpBlockNoteContainer: Pure sync state management. Shows skeleton loader
  while waiting for provider.on('synced'). Only renders BlockNoteEditor
  subcomponent after sync completes.

- BlockNoteEditor: Extracted editor logic with all BlockNote hooks. Only
  mounts when parent confirms sync is done, ensuring useCreateBlockNote
  is called with a Y.Doc that has already received server state.

This leverages React's component lifecycle guarantee: child components only
mount when parent renders them, preventing the hook call until sync is
guaranteed complete.

https://community.openproject.org/wp/69724

akabiru committed:
Extract blocknote attachments to reusable hook fn

akabiru commented:
@copilot open a new pull request to apply changes based on:

* [this feedback](https://github.com/opf/openproject/pull/21415#discussion_r2610452502)
* [this feedback](https://github.com/opf/openproject/pull/21415#discussion_r2610452541)
* [this feedback](https://github.com/opf/openproject/pull/21415#discussion_r2610452602)
* [this feedback](https://github.com/opf/openproject/pull/21415#discussion_r2610452629)

Copilot committed:
Optimize React component performance by memoizing hooks and eliminating unnecessary re-renders (#21417)

brunopagno commented on frontend/src/react/components/BlockNoteEditor.tsx:
```diff
@@ -0,0 +1,133 @@
+  return '#' + Math.floor(Math.random() * 16777215).toString(16).padStart(6, '0');
+}
+
+export function BlockNoteEditor({
```
We should name this component differently, otherwise it clashes with the [BlockNoteEditor](https://github.com/TypeCellOS/BlockNote/blob/main/packages/core/src/editor/BlockNoteEditor.ts) type from `@blocknote/core`.

How about `OpBlockNoteEditor`?

brunopagno reviewed (changes requested):
I think this is beautiful. One tiny change necessary, though.

akabiru committed:
Ensure event listeners are de-registered to prevent stale re-renders

Without clearing old listeners, the editor would get rendered prematurely before the provider
re-synced the empty Y.Doc instance, which triggers an erronoues "empty" doc auto-save, leading to
data loss

akabiru committed:
Rename `OpBlockNoteEditor` to avoid collisions with core blocknote lib

akabiru commented:
https://github.com/opf/openproject/pull/21415/changes/6950e688d1fcab8d843956e8b3f7d3103524330f fixes the data loss from user navigation that was caused by stale `LiveCollaborationManager.onReady` listener functions- that if not cleared, would fire prematurely- such as re-rendering the OpBlockNoteContainer before the provider was synced.

brunopagno reviewed (approved):
🚢 🚢 🚢 🚢
